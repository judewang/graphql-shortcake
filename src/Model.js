/* eslint no-param-reassign: ["error", { "ignorePropertyModificationsFor": ["target"] }] */
import _ from 'lodash';
import fp from 'lodash/fp';
import { invokeNull } from 'thelper';
import withArrayMutator from './withArrayMutator';
import withBuilder from './withBuilder';
import withBatch from './withBatch';
import withCache from './withCache';
import withColumn from './withColumn';
import withDefaultValues from './withDefaultValues';
import withFetcher from './withFetcher';
import withGlobalId from './withGlobalId';
import withIncrementer from './withIncrementer';
import withJSONMutator from './withJSONMutator';
import withLoader from './withLoader';
import withMutator from './withMutator';
import withSearcher from './withSearcher';

class Model {
  static database = null;

  static tableName = null;

  static viewName = null;

  static idAttribute = 'id';

  static hasOperator = false;

  static hasTimestamps = true;

  static softDelete = true;

  static get(target, name) {
    if (this.has(target, name)) {
      return target.get(name);
    }

    return target[name];
  }

  static set(target, name, value) {
    if (this.has(target, name)) {
      target.set(name, value);
      return;
    }

    target[name] = value;
  }

  static get displayName() {
    return this.name;
  }

  static has(target, name) {
    return target.ownKeys.indexOf(name) > -1;
  }

  static ownKeys(target) {
    return target.ownKeys;
  }

  static getOwnPropertyDescriptor(target, name) {
    if (this.has(target, name)) {
      return { configurable: true, enumerable: true, value: target.get(name) };
    }

    return Object.getOwnPropertyDescriptor(target, name);
  }

  static refreshView = _.debounce(() => {
    this.raw(`REFRESH MATERIALIZED VIEW ${this.viewName};`);
  }, 500, { maxWait: 2000 });

  static raw(...args) {
    return this.database.raw(...args);
  }

  static forge(attributes, options) {
    const model = new this({}, options);
    return model.forge(attributes);
  }

  static fromModel(model) {
    return _.get(model, ['nativeId'], model);
  }

  static checkOperator(operator) {
    const { hasOperator } = this;
    if (!hasOperator) return null;
    if (!operator) throw new Error('operator is required');
    return this.fromModel(operator);
  }

  [Symbol.toStringTag] = 'Model';

  [Symbol.toPrimitive](hint) {
    if (hint === 'number') return Number.NaN;
    return this.toString();
  }

  constructor() {
    if (this.initialize) this.initialize();
    const handler = _.pick(this.constructor, [
      'get', 'set', 'has', 'ownKeys', 'getOwnPropertyDescriptor',
    ]);
    return new Proxy(this, handler);
  }

  initialize() {
    this._ = {};
    this.current = {};
    this.previous = {};
  }

  get isNew() {
    const { idAttribute } = this.constructor;
    return !this[idAttribute];
  }

  get values() {
    return { ...this.previous, ...this.current };
  }

  get(paths) {
    const { values } = this;
    return _.get(values, paths, null);
  }

  set(paths, value) {
    _.set(this.current, paths, value);
  }

  valueOf(name) {
    if (name) return _.get(this.current, [name], this.previous[name]);

    const { idAttribute } = this.constructor;
    const id = this.valueOf(idAttribute);
    return id ? String(id) : id;
  }

  toString() {
    const { idAttribute } = this.constructor;
    return this[idAttribute];
  }

  clone(options) {
    const model = new this.constructor({}, options);
    model.current = _.cloneDeep(this.current);
    model.previous = this.previous;
    return model;
  }

  forge(attributes) {
    return invokeNull((attrs) => {
      const values = _.get(attrs, ['previous'], attrs);
      this.current = _.cloneDeep(values);
      this.previous = values;
      return this;
    }, attributes);
  }

  merge(attributes) {
    return invokeNull((attrs) => {
      const { format } = this.constructor;
      const values = format(attrs);
      _.forEach([this.current, this.previous], data => _.assign(data, values));
      return this;
    }, attributes);
  }
}

export default fp.compose(
  withArrayMutator,
  withJSONMutator,
  withIncrementer,
  withSearcher,
  withDefaultValues,
  withCache,
  withLoader,
  withFetcher,
  withMutator,
  withBatch,
  withBuilder,
  withColumn,
  withGlobalId,
)(Model);
