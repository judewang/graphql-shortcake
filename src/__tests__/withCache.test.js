import _ from 'lodash';
import { client } from 'knex';
import database from './database';
import Model, { DataCache, MixedModel, toGlobalId } from '..';

class Fruit extends Model {
  static database = database;

  static tableName = 'cache_fruit';

  static columns = {
    name: { type: String },
  }
}

class Vegetable extends Model {
  static database = database;

  static tableName = 'cache_vegetable';

  static columns = {
    name: { type: String },
  }
}

const Mixed = new MixedModel([Fruit, Vegetable]);

class Airplane extends Model {
  static database = database;

  static tableName = 'cache_airplane';

  static columns = {
    name: { type: String },
    fruit: { type: Fruit },
    food: { type: Mixed },
  }
}

const Cache = new DataCache([Fruit, Airplane, Vegetable]);

describe('Cache', () => {
  beforeAll(async () => {
    await database.schema.dropTableIfExists('cache_fruit');
    await database.schema.createTable('cache_fruit', (table) => {
      table.increments();
      table.string('name');
      table.timestamps();
      table.datetime('deleted_at');
    });

    await new Fruit().insert({ name: 'banana' });
    await new Fruit().insert({ name: 'grape' });
    await new Fruit().insert({ name: 'apple' });

    await database.schema.dropTableIfExists('cache_vegetable');
    await database.schema.createTable('cache_vegetable', (table) => {
      table.increments();
      table.string('name');
      table.timestamps();
      table.datetime('deleted_at');
    });

    await new Vegetable().insert({ name: 'turnip' });

    await database.schema.dropTableIfExists('cache_airplane');
    await database.schema.createTable('cache_airplane', (table) => {
      table.increments();
      table.string('name');
      table.integer('fruit');
      table.string('food');
      table.timestamps();
      table.datetime('deleted_at');
    });

    await new Airplane().insert({ name: 'Airbus A380' });
    await new Airplane().insert({ name: 'Airbus A330', fruit: 1, food: Vegetable.toGlobalId('1') });
    await new Airplane().insert({ name: 'Boeing 777-300ER', fruit: 1, food: Vegetable.toGlobalId('1') });
  });

  describe('update cache', () => {
    it('find', async () => {
      const cache = new Cache();
      const { fruit } = cache;

      expect(_.size(await fruit.fetchAll())).toBe(3);

      await cache.load(Fruit.toGlobalId('1'));

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
    });

    it('insert', async () => {
      const cache = new Cache();
      const { fruit } = cache;

      await fruit.set({ name: 'lemon' }).save();
      await cache.loadFruit(Fruit.toGlobalId(4));
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
    });

    it('update', async () => {
      const cache = new Cache();
      const fruit = await cache.loadFruit(Fruit.toGlobalId(4));
      expect(fruit.name).toBe('lemon');

      await fruit.set({ name: 'lichee' }).save();
      expect(fruit.name).toBe('lichee');

      await expect(cache.loadFruit(Fruit.toGlobalId(4)))
        .resolves.toEqual(expect.objectContaining({ name: 'lichee' }));

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);
    });

    it('destroy', async () => {
      const cache = new Cache();
      const fruit = await cache.loadFruit(Fruit.toGlobalId(4));

      await fruit.destroy();

      await expect(cache.loadFruit(Fruit.toGlobalId(4)))
        .resolves.toBe(null);

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(3);
    });
  });

  describe('load & loadMany', () => {
    it('successfully use cache 1', async () => {
      const cache = new Cache();

      await expect(cache.load(Fruit.toGlobalId('1')))
        .resolves.toEqual(expect.objectContaining({ name: 'banana' }));

      await expect(cache.load(Airplane.toGlobalId('1')))
        .resolves.toEqual(expect.objectContaining({ name: 'Airbus A380' }));

      await expect(Promise.all([
        cache.loadFruit(Fruit.toGlobalId('1')),
        cache.loadAirplane(Airplane.toGlobalId('2')),
        cache.loadMany([Fruit.toGlobalId('3'), Airplane.toGlobalId('1'), Airplane.toGlobalId('4')]),
      ])).resolves.toMatchSnapshot();

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(4);
    });

    it('successfully use cache 2', async () => {
      const cache = new Cache();

      await expect(cache.load(Fruit.toGlobalId('1')))
        .resolves.toEqual(expect.objectContaining({ name: 'banana' }));

      await expect(cache.load(Airplane.toGlobalId('1')))
        .resolves.toEqual(expect.objectContaining({ name: 'Airbus A380' }));

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(2);
    });

    it('successfully loadModel', async () => {
      const cache = new Cache();
      const { fruit } = cache;

      await expect(fruit.loadModel(Airplane, '1'))
        .resolves.toEqual(expect.objectContaining({ name: 'Airbus A380' }));
      expect(client).toHaveBeenCalledTimes(1);

      await expect(fruit.loadModel(Airplane, Airplane.toGlobalId('1')))
        .resolves.toEqual(expect.objectContaining({ name: 'Airbus A380' }));
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(1);
    });

    it('successfully load from columns', async () => {
      const cache = new Cache();
      const airplane = await cache.loadAirplane(2);

      await expect(airplane.fruit)
        .resolves.toEqual(expect.objectContaining({ name: 'banana' }));

      await expect(airplane.food)
        .resolves.toEqual(expect.objectContaining({ name: 'turnip' }));
      expect(client).toHaveBeenCalledTimes(3);

      await expect(airplane.fruit)
        .resolves.toEqual(expect.objectContaining({ name: 'banana' }));
      await expect(airplane.food)
        .resolves.toEqual(expect.objectContaining({ name: 'turnip' }));
      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(3);
    });

    it('check type', async () => {
      const cache = new Cache();

      await expect(cache.load(toGlobalId('tree', '1')))
        .resolves.toBeNull();

      await expect(cache.loadAirplane(Fruit.toGlobalId('10'), TypeError))
        .rejects.toEqual(new TypeError());

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(0);
    });
  });

  describe('DataCache', () => {
    it('clear & clearAll', async () => {
      const cache = new Cache();

      await cache.load(Fruit.toGlobalId('1'));
      await cache.load(Fruit.toGlobalId('1'));
      cache.clear(Fruit.toGlobalId('1'));
      await cache.load(Fruit.toGlobalId('1'));
      await cache.load(Fruit.toGlobalId('2'));
      cache.clearAll();
      await cache.load(Fruit.toGlobalId('1'));

      expect(client).toMatchSnapshot();
      expect(client).toHaveBeenCalledTimes(4);
    });
  });

  it('prime', async () => {
    const cache = new Cache();

    const id = Fruit.toGlobalId('10');
    const fruit = Fruit.forge({ id: '10', name: 'apple' });

    expect(fruit.cache).toBeNull();
    cache.prime(id, fruit);
    expect(fruit.cache).toBe(cache);

    await expect(cache.load(id))
      .resolves.toEqual(expect.objectContaining({ id, name: 'apple' }));

    expect(client).toHaveBeenCalledTimes(0);
  });
});
