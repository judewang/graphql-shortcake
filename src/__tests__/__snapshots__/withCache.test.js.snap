// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cache DataCache clear & clearAll 1`] = `
Array [
  "select * from cache_fruit where id in (1) and deleted_at is null",
  "select * from cache_fruit where id in (1) and deleted_at is null",
  "select * from cache_fruit where id in (2) and deleted_at is null",
  "select * from cache_fruit where id in (1) and deleted_at is null",
]
`;

exports[`Cache load & loadMany check type 1`] = `Array []`;

exports[`Cache load & loadMany successfully load from columns 1`] = `
Array [
  "select * from cache_airplane where id in (2) and deleted_at is null",
  "select * from cache_fruit where id in (1) and deleted_at is null",
  "select * from cache_vegetable where id in (1) and deleted_at is null",
]
`;

exports[`Cache load & loadMany successfully loadModel 1`] = `
Array [
  "select * from cache_airplane where id in (1) and deleted_at is null",
]
`;

exports[`Cache load & loadMany successfully use cache 1 1`] = `
Array [
  Fruit {
    "createdAt": null,
    "deletedAt": null,
    "id": "iN7nPzjYQrQ4JQT3RnINV",
    "name": "banana",
    "updatedAt": null,
  },
  Airplane {
    "createdAt": null,
    "deletedAt": null,
    "food": "iNipQvkLAKkwKj0tp2JExaZhdR",
    "fruit": "iN7nPzjYQrQ4JQT3RnINV",
    "id": "iN8DeCoTu6HHRFcD2CVEWxUp8",
    "name": "Airbus A330",
    "updatedAt": null,
  },
  Array [
    Fruit {
      "createdAt": null,
      "deletedAt": null,
      "id": "iN7nPzjYQrQ4JQT3RnINX",
      "name": "apple",
      "updatedAt": null,
    },
    Airplane {
      "createdAt": null,
      "deletedAt": null,
      "food": null,
      "fruit": null,
      "id": "iN8DeCoTu6HHRFcD2CVEWxUp7",
      "name": "Airbus A380",
      "updatedAt": null,
    },
    null,
  ],
]
`;

exports[`Cache load & loadMany successfully use cache 1 2`] = `
Array [
  "select * from cache_fruit where id in (1) and deleted_at is null",
  "select * from cache_airplane where id in (1) and deleted_at is null",
  "select * from cache_airplane where id in (2, 4) and deleted_at is null",
  "select * from cache_fruit where id in (3) and deleted_at is null",
]
`;

exports[`Cache load & loadMany successfully use cache 2 1`] = `
Array [
  "select * from cache_fruit where id in (1) and deleted_at is null",
  "select * from cache_airplane where id in (1) and deleted_at is null",
]
`;

exports[`Cache update cache destroy 1`] = `
Array [
  "select * from cache_fruit where id in (4) and deleted_at is null",
  "update cache_fruit set deleted_at = DATE where (id = 4) and deleted_at IS NULL returning *",
  "select * from cache_fruit where id in (4) and deleted_at is null",
]
`;

exports[`Cache update cache find 1`] = `
Array [
  "select *, count(*) OVER() AS total_count from cache_fruit where deleted_at IS NULL",
]
`;

exports[`Cache update cache insert 1`] = `
Array [
  "insert into cache_fruit (created_at, name, updated_at) values (DATE, lemon, DATE) returning *",
]
`;

exports[`Cache update cache update 1`] = `
Array [
  "select * from cache_fruit where id in (4) and deleted_at is null",
  "update cache_fruit set name = lichee, updated_at = DATE where (id = 4) and deleted_at IS NULL returning *",
]
`;
